### ISSUE TITLE
Launch an external process in the build window

### ISSUE THREAD

**Issue by hadley** [rstudio/rstudioapi#17]
It would be nice if we could do this ourselves, for other long-running check processes (e.g. the revdep checks)

**Comment by hadley** [comment]
This would also be a nice option for add-ins.

**Comment by jjallaire** [comment]
We could also handle this via the terminal API (i.e. launch the process in
a terminal rather than in the build pane)

**Comment by kevinushey** [comment]
I think @gtritchie is handling the terminal-specific stuff here, but it would definitely be nice to allow users to run arbitrary R / shell commands that stream output into the build pane. Hopefully we can meet up and chat in person this work week about what we should do here exactly.

**Comment by kevinushey** [comment]
I think, for the scope of this issue, we want to make it possible to launch processes in the terminal -- fortunately, @gtritchie is working on exactly that.

The larger feature of allowing packages / functions to take over the build pane is something we can consider separately, I think.

**Comment by gtritchie** [comment]
Correct, it's next on my list after my current task. I already did much of the necessary plumbing in RStudio itself. Idea is that you can launch a command, specify environment and current working directory, see the output in a read-only terminal tab, and query the result (both the terminal buffer and the process' exit code).

**Comment by jonkeane** [comment]
Is there any chance this code is on a branch (of rstudioapi) or beta release (of rstudio) somewhere? I would be happy to take a stab at making it happen in rstudioapi (and submit a PR for it). This would be super helpful for a project I'm working on.

**Comment by vnijs** [comment]
I'm wondering if this would allow also `shiny::runGadget` to work in a separate R process but still show in Rstudio Viewer?

See also question posted on the shiny google group a while back: https://groups.google.com/forum/#!searchin/shiny-discuss/vincent%7Csort:date/shiny-discuss/aQfKp_pIHow/tr6TTX_vBQAJ

cc @jcheng5 @daattali

**Comment by jcheng5** [comment]
@vnijs I think it's unrelated. This seemed to work?

```r
# Launch Shiny app in another process, without blocking
callr::r_bg(function() { shiny::runExample("01_hello", port = 3131, launch.browser = FALSE) })
# Launch viewer
getOption("viewer")("http://localhost:3131/")
```

**Comment by daattali** [comment]
With gadgets, they often return values on exit, so I might actually want to capture the result of runGadget. 
is this possible? (I did not do my homework and didn't spend much time reading through `callr` documentation)

**Comment by jcheng5** [comment]
@daattali Yes that should be fine as long as it's a normal-ish R object you're returning (no external pointers, open sockets, etc.). The call to `r_bg` gives you back an object that you can query for the status and returned object. My upcoming promise library will make it fairly straightforward to deal with such objects.

**Comment by sz-cgt** [comment]
> This would also be a nice option for add-ins.

Is this now feasible for addins?  Or still on a list somewhere?

**Comment by netique** [comment]
I am developing a package that aims to help users to manage their build scripts (their goal is to source several scripts in given order and render a RMarkdown in the very end, to give an example). I am able to run the selected rule in the terminal, but it would be a lot nicer to show the output in the respective build pane.  However, solving #214 may be better approach since Build pane should be preserved for build-related stuff only in my view.
